### PR TITLE
[python] Remove use of generic python shebang, update CI

### LINKED ISSUE #10465: [ci] [azure] Azure workers don't pass the test suite due to missing `python`
Since today Azure workers have been failing to pass the test suite:
```
/usr/bin/env: 'python': No such file or directory
```

### PR DESCRIPTION
**Pull request by ejgallego**
Fixes #10465

Following discussion we don't allow a generic `/usr/bin/python`
shebang anymore. We thus move all the scripts [but one] to python3,
and add python3 to the Azure base environment.

Unfortunately we still depend on python2 for the update-compat.sh
script, see #10491

We thus have to complement #10467 adding python2 back to Nix,
until #10491 is fixed.

### CODE REVIEW

**Review by Zimmi48 — COMMENTED** [review]
Huh! We are not supposed to depend on python2, are we? @JasonGross?

**Comment by ejgallego** [comment]
> Huh! We are not supposed to depend on python2, are we? @JasonGross?

Sure we do, check the nix file.

**Comment by Zimmi48** [comment]
The Nix file was never updated after we removed the dependency on python2 in #7990. I wanted to clean it and never did it, and I obviously should have if you are telling me that we accidentally reintroduced such a dependency. I would have never willingly merged a PR that did.

**Comment by SkySkimmer** [comment]
The failure isn't caused by a dependency on python2, just look at the error message. I don't know why you think this would fix it.

**Comment by ejgallego** [comment]
@Zimmi48 certainly #7790 seems bogus at various levels, as `/usr/bin/env python` is bound to python2 by default, for example in Ubuntu 19.04

**Comment by Zimmi48** [comment]
Please don't insult contributors' work for no reason. Nowadays, some distributions bind `python` to `python2` and others to `python3`. Using a generic shebang is just supposed to carry the meaning that you can use either because you made your script compatible, which is what this PR did.

**Comment by ejgallego** [comment]
> Please don't insult contributors' work for no reason.

@Zimmi48 calm down I'm not insulting anyone, indeed the PR does look incorrect to me, for one, there are a few uses of `python2` in the codebase; I could be wrong of course, but there seem to be more problems. Using the generic shebang is also a serious problems these days [python experts tell me]

**Comment by ejgallego** [comment]
p.s: [bogus -> incorrect] edited in my previous comment, "bogus" for me is a playful informal word but my apologies if anyone found it "insulting"

**Comment by ejgallego** [comment]
> The failure isn't caused by a dependency on python2, just look at the error message. I don't know why you think this would fix it.

I'm not sure what you mean, but indeed this PR fixes the problem. Intended or not, we still depend on python2 , as least all the testing for these scripts has been done using python2 even after #7990

**Comment by SkySkimmer** [comment]
>The failure isn't caused by a dependency on python2, just look at the error message. I don't know why you think this would fix it.

Apparently I was wrong. I have decreasingly less of a clue what's going on.

**Comment by ejgallego** [comment]
> I have decreasingly less of a clue what's going on.

What happened is that cygwin deprecated the generic python alias, now you need to use python2, which is what we were using before. Of course it'd be great to complete #7990 and move everything to P3, but IMHO that should go in a different PR.

**Comment by Zimmi48** [comment]
I didn't say you insulted anyone but the *work* (that was done by an external contributor), by making false claims (or at the very least these less than convincing claims).

>Using the generic shebang is also a serious problems these days [python experts tell me]

If so, then the same would apply to using generic shell shebangs. It carries pretty much the same kind of meaning, and it has the same problems: we only test them with whatever shell is installed (probably bash). Anyway, if you really wanted to pin shebangs, then you should pin them to `python3`.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
>Of course it'd be great to complete #7990 and move everything to P3, but IMHO that should go in a different PR.

Where do you see any hint that this work needs anything more to be complete?

If you don't want to use the generic alias, that's fine, but then use `python3` everywhere, not `python2`.

**Comment by ejgallego** [comment]
> Where do you see any hint that this work needs anything more to be complete?

That PR forgot to do 3 things:
- switch the docked image so the timing tests are run using python3 [they are run using python2 as of today]
- switch the azure image so the timing tests are run using python3 [they are run using python2 as of today]
- remove python2 from the nix files.

I disagree with your request for changes, what this PR does is to document what's being actually tested. Please update the shebangs in a different PR, thanks.

**Comment by ejgallego** [comment]
> If so, then the same would apply to using generic shell shebangs

The situation with shell shebangs is quite different from the one with python shebangs, even tho you have people proposing to remove `/usr/bin/sh` as cygwin did with `/usr/bin/python`

**Comment by ejgallego** [comment]
> I didn't say you insulted anyone but the _work_ (that was done by an external contributor), by making false claims (or at the very least these less than convincing claims).

IMVHO you are a bit off the line here, first, in a development / scientific context criticizing a work is not an insult, even if my critic turns out to be wrong. I am human and I do have the right to be wrong and I am actually wrong most of the time.

Also, I have a hard time following how who did the work is relevant if your comment was about the work itself.

About the claims, I'm unsure why they are not convincing, it seems to me that the 3 problems I pointed out above do hold, but I could be wrong of course.

**Comment by Zimmi48** [comment]
I'll stop debating on the question of criticizing the work, and accept that you have reasons to think it was incomplete.

However, I maintain my request for changes because changing the shebangs like you did is not just a matter of documenting the preexisting situation, it is a breaking change that means that suddenly all developers will need to have python2 installed in their machines, and typically I don't, and once #10467 is merged, I really won't, even within a `nix-shell` environment. And there is no reason why the fix you are proposing to the Windows testing infrastructure should affect the ability of merging #10467.

**Comment by ejgallego** [comment]
Indeed that seems like a more sensible approach; updated. Test suite passes locally, let's see if the rest still works.

**Comment by ejgallego** [comment]
Some confusion here came from the fact than in the debian world, python is always python2 TTBOMK.

**Comment by ejgallego** [comment]
So `update-compat.sh` is broken in linux for python3, I see no other short-term option than reverting the python3 removal for that script.

**Comment by ejgallego** [comment]
> So `update-compat.sh` is broken in linux for python3, I see no other short-term option than reverting the python3 removal for that script.

Ok, that seems to have worked, @Zimmi48 is that tradeoff acceptable to you?

**Comment by SkySkimmer** [comment]
>So update-compat.sh is broken in linux for python3, I see no other short-term option than reverting the python3 removal for that script.

Do you know why it works in https://github.com/coq/coq/pull/10467 ?

**Comment by ejgallego** [comment]
> Do you know why it works in #10467 ?

I have no idea, likely differences with env locale variables?

**Comment by Zimmi48** [comment]
I'd like to get the help of @JasonGross to understand what's going on. In the meantime, I propose an alternative temporary fix which is to create a symlink `python` to `python2` in Azure to mimic what cygwin doesn't do anymore. Even assuming that it is indeed recommended today in the Python world to always use `python2` or `python3` (and I kind of doubt that this is consensual), I think it was probably a bad move from Cygwin to remove `python` entirely, and it has probably broken lots of software.

**Comment by SkySkimmer** [comment]
They didn't just change it to point to python3?

**Comment by ejgallego** [comment]
> I propose an alternative temporary fix which is to create a symlink `python` to `python2` in Azure to mimic what cygwin doesn't do anymore.

What's the point of that? Shouldn't we instead document the actual status-quo w.r.t. testing?

The script is run in python2 not only in Azure, but in the Docker containers.

IMHO this PR should not fix the Python code, it is just improving what was an underspecification in the CI that created many problems, and fixing our CI system.

Already we are changing the python version for the timing scripts at your request, thing which I am pretty uncomfortable to do as this should be done by Python people not by accident by me.

**Comment by ejgallego** [comment]
> They didn't just change it to point to python3?

I dunno, but I am pretty certain that we don't want to do that kind of hack and create more problems.

**Comment by ejgallego** [comment]
> What's the point of that?

Oh I see, python2 is not more on Nix. I insisted that such removal was premature as there were at least 4 scripts being tested on python2. We were lucky in that we only discovered problems in one of them.

As of now, this PR + reverting #10467 is the right thing to do.

**Comment by ejgallego** [comment]
Anyways folks IMHO we should not get very entangled in trying to solve this issue. Let me propose a plan of action that should hopefully solve the situation.

## Current situation

We have a discrepancy on the python version used for running scripts that declare a "generic" `/usr/bin/python` shebang:
- in Nix they were tested using python3
- in Docker images and Azure they were tested using python2
- the `update-compat.py` python script doesn't work with python3 in Docker

## Analysis

Using the generic shebang seems like a bad idea, particularly as some environments have deprecated it.

## Plan of action

1. make all scripts use a versioned shebang.
2. make all the environments install python2 and python3
3. solve the remaining problem with `update-compat.sh`, remove python2 from envs

Steps 1 and 2 could be done in this PR, step 3 would be its own PR, by the Python maintainers.

WDYT?

**Comment by Zimmi48** [comment]
Since the only urgency is to fix the failing Windows test, I don't get what you have against my proposal as *a temporary solution*. Again, I would view pinning to python2 as a regression. So this proposed solution avoids this issue and give some time to knowledgeable people to understand and fix the underlying issue. I don't see why some unreasonable decision from the Cygwin maintainers should have any impact on our Linux users/developers.

**Comment by ejgallego** [comment]
> I don't get what you have against my proposal as _a temporary solution_.

I think that your proposal has many drawbacks, including doing a non-supported hack that we don't know what would it break. Moreover, we don't gain a lot by avoiding the pinning, as this is the status-quo on master.

> Again, I would view pinning to python2 as a regression.

I fail to see why is a regression, as testing is done today already with python2, and #10467 was merged without taking into consideration CI.


> So this proposed solution avoids this issue and give some time to knowledgeable people to understand and fix the underlying issue.

So does mine, without doing any hack.

> I don't see why some unreasonable decision from the Cygwin maintainers should have any impact on our Linux users/developers.

IMO we have no choice other than go with that. I don't get what impact my proposal would have on Linux users / developers, I see it as zero impact, however maintaining the current undefined behavior does have a negative impact.

**Comment by JasonGross** [comment]
I think I'm with @ejgallego on this one, given that the only script we're pinning to python2 is the update-compat script, which apparently does not work with python3, which means that developers without python2 already cannot run it, and so pinning it to python2 does not seem to me to be a breaking change.

As for why it doesn't work with python3, I'm not sure, and I don't have time to dig into it right now.  What's the error message?  Is the issue Unicode characters / bytes?

**Comment by ejgallego** [comment]
> Is the issue Unicode characters / bytes?

Looks like it @JasonGross , for some reason it only failed on Linux / Ubuntu.

**Comment by ejgallego** [comment]
Here's the log @JasonGross 
```
Traceback (most recent call last):
  File "dev/tools/update-compat.py", line 481, in <module>
    update_compat_notations(known_versions, new_versions, **args)
  File "dev/tools/update-compat.py", line 413, in update_compat_notations
    with open(os.path.join(root, fname), 'r') as f: contents = f.read()
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1390: ordinal not in range(128)
==========> FAILURE <==========
    tools/update-compat/run.sh...Error!
```
https://gitlab.com/coq/coq/-/jobs/244148789/artifacts/browse/test-suite/logs/tools/

**Comment by vbgl** [comment]
Is the `update-compate.py` python2 program run at some point in CI?

**Comment by ejgallego** [comment]
> Is the `update-compate.py` python2 program run at some point in CI?

Yes.

**Comment by vbgl** [comment]
When?

**Comment by ejgallego** [comment]
> When?

IIRC the test suite has a test for it.

**Comment by vbgl** [comment]
Then how come that job succeeded: https://gitlab.com/coq/coq/-/jobs/244075110? The log shows the following lines:

> TEST      tools/update-compat
> …
> NO FAILURES

There was no python2 available there…

**Comment by ejgallego** [comment]
> There was no python2 available there…

I have no idea what the Nix setup is, in Debian it does fail, as pointed out above.

**Comment by vbgl** [comment]
> I have no idea what the Nix setup is

Therefore, why do you try to mess up with it in this PR?

**Comment by ejgallego** [comment]
The thing is that I don't see a big problem in pinning python2 for while more, while the remaining script and the missing test suite entry are fixed.

**Comment by ejgallego** [comment]
> Therefore, why do you try to mess up with it in this PR?

Because the guilty script now explicitly uses a /usr/bin/python2 shebang.

**Comment by ejgallego** [comment]
So it seems more sensible to me document the status quo by always using versioned shebangs, then fix the script later. The other choice is fixing the script first then pinning to python3 only.

**Comment by ejgallego** [comment]
Real underlying problem here tho is that the windows CI infrastructure is setup in a non reproducible way, but that's a big task to tackle I'm much afraid.

**Comment by Zimmi48** [comment]
The difference could also have to do with the version of Python3. In Nix, it is 3.7.3. According to https://packages.ubuntu.com/bionic/python3, the version on our Ubuntu Docker image is 3.6.5. Maybe this version has a bug...

**Comment by ejgallego** [comment]
> Maybe this version has a bug...

Could be, I dunno, I'm not a Python expert.

Anyways IMHO it is time to process this, given that @JasonGross agrees in that pinning python2 is a better overall strategy, can we proceed to merge?

**Comment by Zimmi48** [comment]
I'm still not convinced that this is the right way to proceed, but I will let @vbgl decide.

**Comment by ejgallego** [comment]
> I'm still not convinced that this is the right way to proceed

Why? Théo I don't understand where you see the problem, maybe I'm missing something, care to explain more?

**Comment by ejgallego** [comment]
Chat on Gitter?

**Review by Zimmi48 — APPROVED** [review]
Following this PR, we won't any longer accept scripts with generic python shebangs. This is the only way to ensure that we don't accidentally reintroduce a dependency on Python2. And I hope we can upgrade the update script to Python3 sooner than later.

Reminder: the Python2 clock is ticking https://pythonclock.org/

**Comment by ejgallego** [comment]
Thanks to all, PR title and commit message updated. Hopefully this should be good until we can fix #10491

**Comment by ejgallego** [comment]
@SkySkimmer can you assign? Thanks!

File: default.nix
Comment (SkySkimmer) [inline default.nix]: Isn't it .py?

Comment (ejgallego) [reply default.nix]: Yup.

**Comment by ejgallego** [comment]
@SkySkimmer this is ready to go.

**Comment by ejgallego** [comment]
Merci !

**Pull request merged**
